<a href="https://colab.research.google.com/github/karsarobert/ChatGpt_2024/blob/main/07/ChatGPT2024_07.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Chat GPT és más nagy nyelvi modellek alkalmazása
##PTE Gépi tanulás
###7. gyakorlat: OCR
2024. március 18.

##Alkalmazás az OCR-technológiával:
A felhasználó képet (pl. számlát) tölt fel OCR-feldolgozásra
Mesterséges intelligencia rendezi a kinyerhető szöveget
Az adatokból például automatikusan táblázatba kerülnek a költségek

In [1]:
!gdown 1GesvS7nhgHK3eke-t2RsQYGXGSVy_2g1 #adatok beszerzése
!gdown 1OcuK4c6UdoqGGSRPEn3Zxip8EfvhBXQU
!gdown 1RdBidgJM4RxdgKtRjGtgZ1TqwV7g5i1y

Downloading...
From: https://drive.google.com/uc?id=1GesvS7nhgHK3eke-t2RsQYGXGSVy_2g1
To: /content/Aldi.jpg
100% 253k/253k [00:00<00:00, 52.3MB/s]
Downloading...
From: https://drive.google.com/uc?id=1OcuK4c6UdoqGGSRPEn3Zxip8EfvhBXQU
To: /content/szamla1.jpg
100% 126k/126k [00:00<00:00, 71.7MB/s]
Downloading...
From: https://drive.google.com/uc?id=1RdBidgJM4RxdgKtRjGtgZ1TqwV7g5i1y
To: /content/szamla.jpg
100% 161k/161k [00:00<00:00, 49.7MB/s]


In [2]:
!apt-get install tesseract-ocr
!pip install pytesseract pillow
!sudo apt-get install tesseract-ocr-hun -y

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  tesseract-ocr-eng tesseract-ocr-osd
The following NEW packages will be installed:
  tesseract-ocr tesseract-ocr-eng tesseract-ocr-osd
0 upgraded, 3 newly installed, 0 to remove and 38 not upgraded.
Need to get 4,816 kB of archives.
After this operation, 15.6 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 tesseract-ocr-eng all 1:4.00~git30-7274cfa-1.1 [1,591 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 tesseract-ocr-osd all 1:4.00~git30-7274cfa-1.1 [2,990 kB]
Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 tesseract-ocr amd64 4.1.1-2.1build1 [236 kB]
Fetched 4,816 kB in 1s (6,171 kB/s)
Selecting previously unselected package tesseract-ocr-eng.
(Reading database ... 121752 files and directories currently installed.)
Preparing to unpack .../tesseract-ocr-

In [6]:
import requests
from PIL import Image
from io import BytesIO

url = "https://www.infocentrum.hu/wp-content/uploads/help_images/elolegszamla_kep.png"

response = requests.get(url)
image_bytes = response.content

kep = Image.open(BytesIO(image_bytes))

Prompt: készíts egy python szkriptet colabhoz ami beolvas egy képet majd az tesseract segtítségével felismeri a szövegeket, táblázatokat és kiírja azokat.

In [7]:
from PIL import Image
import pytesseract
import io

# Beolvasandó kép elérési útja
kep_eleresi_ut = "Aldi.jpg"

# Kép betöltése
kep = Image.open(kep_eleresi_ut)

# Szöveg kinyerése
szoveg = pytesseract.image_to_string(kep,lang="hun")

# Táblázat kinyerése
tabulat = pytesseract.image_to_data(kep, lang="hun", output_type=pytesseract.Output.DICT)

# Kiíratás
print("Szöveg:")
print(szoveg)

#print("Táblázat:")
#for sor in tabulat['text']:
#    print(sor)


Szöveg:
   
  
  
  
    
  
 

Adi Magyarország Élelmiszer Bt.
2051 Biatorbágy
Mészárosok útja 2.

102. sz, üzlet
7634 ker
Nyugati Ipari
ADÓSZÁM: 22934663- 3. 44

— WIN —

B00 ME Ek 1799 1.52

5 DB x 75 Ft/DB 375
(00 sza nö.olaj . 1299 399
Ü S; Hi 999
BŰN Félzs. túró 2509 449
500 Félzs. túró 2509 449
B00 Tejföl 207 4509 469
T 499
619
619
49 Ft/DB 1347
00. Ananász 9809 999
COD Ananász 9809 999
C00 Almapüré a 99
9 299
SZ 9 399
21 699
CO P 4 429
C00 Ananász 5809 944
C00 Zabpehely 9009 399
(00 ketchup 9999 Üz
Bo ki ő . ex. Sz, 790m! 2799
OVáSZOS CIPÓ
3 DB x 499 Ft/DB 1497
C00 Brokkoli 4509 649
BD Fr .camemb. 2909 4 180
an 8 tea 1do 299
urgonya Pp.
. 4107 1 Tára
728
1 459
1798
7699
21.269 Ft

 

21 263 Ft
21 263 Ft



In [8]:
tabulat

{'level': [1,
  2,
  3,
  4,
  5,
  4,
  5,
  4,
  5,
  4,
  5,
  4,
  5,
  5,
  4,
  5,
  4,
  5,
  2,
  3,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  5,
  2,
  3,
  4,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  2,
  3,
  4,
  5,
  5,
  5,
  2,
  3,
  4,
  5,
  5,
  5,
  5,
  5,
  2,
  3,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  4,
  5,
  4,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
 

##Professzionális megoldás a problémára: LayoutLM

In [18]:
print(pytesseract.image_to_data(kep, lang="hun", output_type=pytesseract.Output.DICT))

{'level': [1, 2, 3, 4, 5, 2, 3, 4, 5, 4, 5, 5, 2, 3, 4, 5, 5, 2, 3, 4, 5, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 4, 5, 5, 5, 4, 5, 5, 5, 5, 5, 2, 3, 4, 5, 5, 4, 5, 5, 5, 4, 5, 4, 5, 3, 4, 5, 5, 4, 5, 2, 3, 4, 5, 4, 5, 5, 2, 3, 4, 5, 5, 4, 5, 5, 5, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 5, 5, 5, 5, 5, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 5, 5, 5, 5, 4, 5, 5, 4, 5, 5, 5, 5, 5, 5, 4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 5, 5, 5, 5, 5, 5, 4, 5, 5, 5, 5, 5, 5, 5, 4, 5, 5, 5, 4, 5, 5, 5, 5, 5, 4, 5, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 5, 5, 2, 3, 4, 5, 2, 3, 4, 5, 5, 5, 5, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 5, 5, 5, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 5, 5, 5, 5, 5, 4, 5,

Prompt: A következő számla szövegben keresd meg szolgáltató adatait és a fizetendő összeget készíts belőle JSON objektumot:
szöveg:

  
    
  


Adi Magyarország Élelmiszer Bt.
2051 Biatorbágy
Mészárosok útja 2.

102. sz, üzlet
7634 ker
Nyugati Ipari
ADÓSZÁM: 22934663- 3. 44

— WIN —

B00 ME Ek 1799 1.52

5 DB x 75 Ft/DB 375
(00 sza nö.olaj . 1299 399
Ü S; Hi 999
BŰN Félzs. túró 2509 449
500 Félzs. túró 2509 449
B00 Tejföl 207 4509 469
T 499
619
619
49 Ft/DB 1347
00. Ananász 9809 999
COD Ananász 9809 999
C00 Almapüré a 99
9 299
SZ 9 399
21 699
CO P 4 429
C00 Ananász 5809 944
C00 Zabpehely 9009 399
(00 ketchup 9999 Üz
Bo ki ő . ex. Sz, 790m! 2799
OVáSZOS CIPÓ
3 DB x 499 Ft/DB 1497
C00 Brokkoli 4509 649
BD Fr .camemb. 2909 4 180
an 8 tea 1do 299
urgonya Pp.
. 4107 1 Tára
728
1 459
1798
7699
21.269 Ft



21 263 Ft
21 263 Ft


Kimenet:
{
  "szolgaltato": {
    "nev": "Adi Magyarország Élelmiszer Bt.",
    "cim": "2051 Biatorbágy, Mészárosok útja 2., 102. sz, üzlet, 7634 ker",
    "adoszam": "22934663-3-44"
  },
  "fizetendo_osszeg": 21263
}


In [9]:
{
  "szolgaltato": {
    "nev": "Adi Magyarország Élelmiszer Bt.",
    "cim": "2051 Biatorbágy, Mészárosok útja 2., 102. sz, üzlet, 7634 ker",
    "adoszam": "22934663-3-44"
  },
  "fizetendo_osszeg": 21263
}


{'szolgaltato': {'nev': 'Adi Magyarország Élelmiszer Bt.',
  'cim': '2051 Biatorbágy, Mészárosok útja 2., 102. sz, üzlet, 7634 ker',
  'adoszam': '22934663-3-44'},
 'fizetendo_osszeg': 21263}

#Feladat:

![img](https://www.infocentrum.hu/wp-content/uploads/help_images/elolegszamla_kep.png)

Elad

In [10]:
import requests
from io import BytesIO

url = "https://www.infocentrum.hu/wp-content/uploads/help_images/elolegszamla_kep.png"

response = requests.get(url)
image_bytes = response.content

kep = Image.open(BytesIO(image_bytes))

In [11]:
szoveg = pytesseract.image_to_string(kep,lang="hun")

# Táblázat kinyerése
tabulat = pytesseract.image_to_data(kep, lang="hun", output_type=pytesseract.Output.DICT)

# Kiíratás
#print("Szöveg:")
#print(szoveg)

print("Táblázat:")
for sor in tabulat['text']:
    print(sor)

Táblázat:




 



infocentrum

szoftver
stúdió



Előleg
számla



Sorszám:
000007/2015



 



 



Szállító

Infocentrum
Szoftver
Stúdió

7400
Kaposvár,
Vár
u.
12-14.



Adószám:
12345678-1-12

Közösségi
adószám
HU12345678

Bankszámlaszám:

tel:06-30-916-8888


fax
821312-391

wwrwinfocentrumhu



Vevő

TesztKereskedelmi
Kft.



Adószám:
12121212-1-21

Közösségi
adószám
HU12121212

Bankszámlaszám:



 



1023
Budapest
II.
kerület,
Madár
u.
55.



 



 



 



 



 



 



 



 



 



 



 



 



 



 



 



 



 



 



 



Megrendelés
száma:
Fizetés
módja:
Telj.
időpontja:
Számla
kelte:
Fiz.
határidő;

Átutalás
2015.06.12.
2015.06.12.
2015.06.12.

Egyéb
adatok:

orolási
szám,
Megnevezés
e.
ÁT]
Átaérté

VTSZ/
Besorolási
szám,
Megi
ME]
Menny.
[
ágmélken
té]
Áraérték

ELŐLEG
ri
a]
50
000
50000]
2796
13500
63
500]

4
lioenszes
hálózatos
számlázó
praram
előlege.

Ji
dési
ök
tsdlé
Számlaérték:
50.000
Ft

Áfa:
13500
Ft

27
50000
Ft
13500Ft
63500Ft

Fizetendő:
$3500Ft



 

In [12]:
import yfinance as yf
import pandas as pd
# Define the ticker symbols
tickers = ['AAPL', 'GOOG', 'MSFT']
# Define the start and end dates
start_date = '2020-01-01'
end_date = '2022-12-31'
# Create an empty DataFrame to store the data
data = pd.DataFrame()
# Download the data
for ticker in tickers:
 df = yf.download(ticker, start=start_date, end=end_date, interval='1mo')
 df['Ticker'] = ticker # Add a column with the ticker symbol
 data = pd.concat([data, df])
# Reset the index
data.reset_index(inplace=True)
# Save the data to a CSV file
data.to_csv('stock_data.csv', index=False)

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


In [15]:
!pip install openai

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 257.5/257.5 kB 5.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.6/75.6 kB 7.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.8/77.8 kB 7.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 6.3 MB/s eta 0:00:00


In [16]:
import os
from openai import OpenAI
client = OpenAI(api_key= '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')

import pandas as pd

csv_string = data[:2].to_string(index=False)


input_text = f"adok egy sor céginformációt a következő sorrendben Cég,Ár-arányos nyereség,Ár-arányos könyv szerinti érték,Saját tőke megtérülése %,Adósság a saját tőkéhez,Aktuális mutató,Bruttó árrés %,elemezze az egyes cégek mutatóit, hivatkozzon a cég nevére, és írjon egy tömör választ magyarul.\n\n{csv_string}"
response = client.completions.create(
    model="gpt-3.5-turbo-instruct",  # Specifying GPT-3.5-turbo engine
    prompt=input_text,
    max_tokens=300  # Limiting the length of the generated text
)
generated_text = response.choices[0].text.strip()
print("GPT-3.5-turbo Cégelemzés:", generated_text)

GPT-3.5-turbo Cégelemzés: Apple: Cég,1.03,480.67,66.49,1.37,74.06%, 6.03% - Az Apple ár-arányos nyeresége alacsony és a könyv szerinti érték magas, ami jó jel, a saját tőke megtérülése magas, ami azt mutatja, hogy a cég hatékonyan kezeli a tőkéjét. Az adósság a saját tőkéhez viszonylag alacsony, ami jó hír a befektetők számára. Az aktuális mutató magas, ami az összes eszköz és az összes kötelezettség viszonyát mutatja, ami azt jelenti, hogy az Apple-nél kevés likviditási probléma van. A bruttó árrés magas, ami azt jelenti, hogy az átváltási ár és a nettó érték között jelentős különbségek vannak. Ezek az adatok arra utalnak, hogy az Apple jól teljesít és stabil gazdasági helyzetben van.
